modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4SizeHeaderFrameDecoder.java

```diff
@@ -46,9 +46,11 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
                     if (messageLengthWithHeader > in.readableBytes()) {
                         continueDecode = false;
                     } else {
-                        final ByteBuf message = in.retainedSlice(in.readerIndex() + HEADER_SIZE, messageLength);
+                        final int readerIndex = in.readerIndex();
+                        final ByteBuf message = in.retainedSlice(readerIndex + HEADER_SIZE, messageLength);
                         out.add(message);
-                        in.readerIndex(in.readerIndex() + messageLengthWithHeader);
+                        in.readerIndex(readerIndex + messageLengthWithHeader);
+                        continueDecode = in.readableBytes() >= HEADER_SIZE;
```

Comment:
simpler to just write:

```
while (in.readableBytes() >= HEADER_SIZE) {
    int messageLength = TcpTransport.readMessageLength(Netty4Utils.toBytesReference(in));
    if (messageLength == -1) {
        break;
    } else {
        int messageLengthWithHeader = messageLength + HEADER_SIZE;
        // If the message length is greater than the network bytes available, we have not read a complete frame.
        if (messageLengthWithHeader > in.readableBytes()) {
            break;
        } else {
            final ByteBuf message = in.retainedSlice(in.readerIndex() + HEADER_SIZE, messageLength);
            out.add(message);
            in.readerIndex(in.readerIndex() + messageLengthWithHeader);
        }
    }
}
```

Reply:
Right sorry about that, fixed in 9d1f6a6